content/crutchcorn/posts/why-is-css-in-js-slow/index.md

```diff
@@ -0,0 +1,261 @@
+
+![The HTML file loads a JS bundle from the <body>, which generates the markup from <App/>, which in turn generates a <style> tag. This causes the following waterfall chain: HTML is parsed and displayed. Then JS is detected in the HTML, parsed, and displayed in the HTML (without <App/>'s markup). Then the framework's JS starts, generates the markup and displays in the browser. Finally, the CSS is generated, parsed, and displayed with everything else.](./framework_css_in_js_handling.png)
+
+# Compiled CSS-in-JS
+
+> "But not all CSS-in-JS solutions are slow!"
+
+This is true. If you're using solutions like [StyleX](https://stylexjs.com/), [PandaCSS](https://panda-css.com/), or [Vanilla Extract](https://vanilla-extract.style/), you're not likely to run into these same performance problems of JS to CSS waterfalling.
+
+However, the reason these frameworks aren't slow are because they sidestep the "JavaScript" part of the problem almost entirely.
+
+See, these tools all utilize a compiler of some kind to compile away from runtime JavaScript into compiled CSS. For example, this PandaCSS code:
+
+```jsx
+import { css } from './styled-system/css'
+ 
+export function App() {
+  return <div className={css({ bg: 'red.400' })} />
+}
+```
+
+Might compile to something like this **during your site's compilation, on your developer machine**:
+
+
+```html
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        <style>
+        .bg_red\.400 {
+            background: var(--colors-red-400);
+        }
+        </style>
+    </head>
+    <body>
+        <!-- ... -->
+    </body>
+</html>
+```
+
+```jsx
+// App.jsx
+import { css } from './styled-system/css'
+ 
+export function App() {
+  return <div className={"bg_red.400"} />
+}
+```
+
+This takes our problem back from this:
+
+![Our complex Framework CSS-in-JS handling before](./framework_css_in_js_handling.png)
+
+To the original problem space like this:
+
+![Our more straightforward loading of CSS from vanilla HTML from before](./html_css_handling.png)
+
+> **Not just these tools:**
+>
+> Earlier I mentioned a list of CSS-in-JS libraries that were slow and ones that compiled to CSS. While this is _generally_ true, it's not always so black and white.
+>
+> After all, some libraries have plugins or [SSR](https://playfulprogramming.com/posts/what-is-ssr-and-ssg) compatibility to either reduce or solve the problems with CSS-in-JS as outlined.
+
+# Conclusion
+
+Hopefully this has been an insightful look into how different CSS-in-JS solutions work under-the-hood and why you might or might not want to select one  library over another.
+
+Want to learn more about styling options in your framework? Check out [the second book in The Framework Field Guide series](https://framework.guide#ecosystem-title).
+
+[![The Framework Field Guide, a free and pratical way to learn Angular, React & Vue all at once.](./ffg_ecosystem.png)](https://framework.guide#ecosystem-title)
```

Comment:
Just had a chance to review this right now

I think this should say `practical` instead of `pratical`.

Reply:
Sure!